va/va_prot.h

```diff
@@ -0,0 +1,361 @@
+ *
+ * vaDestroyBuffer(va_dpy, buffer);
+ * \endcode
+ *
+ * \section api_prot_exec TEE communication via vaProtectedSessionTeeExec()
+ *
+ * A function for communicating with TEE directly (Trusted Execution Environment). It needs to
+ * create a VA buffer first with type VAProtectedSessionTeeExecBufferType, the structure of
+ * this type is IHV's implementation specific. Usually it will have a function_id to specify
+ * which function to execute. The following pseudo-code demonstrates getting session id via
+ * vaProtectedSessionTeeExec() as an example.
+ *
+ * \code
+ * uint16_t version = 0xFF;
+ * VABufferID buffer;
+ * VAProtectedSessionTeeExecBuffer tee_exec_buff = {0};
+ *
+ * tee_exec_buff.function_id = GET_TEE_VERSION
+ * tee_exec_buff.input.input_data = nullptr;
+ * tee_exec_buff.input.data_size = 0;
+ * tee_exec_buff.output.output_data = &version;
+ * tee_exec_buff.output.max_data_size = (uint16_t);
+ * va_status = vaCreateBuffer(
+ *                 va_dpy,
+ *                 protected_session,
+ *                 (VABufferType) VAProtectedSessionTeeExecBufferType,
+ *                 sizeof(tee_exec_buff),
+ *                 1,
+ *                 &tee_exec_buff,
+ *                 &buffer);
+ *
+ * va_status = vaProtectedSessionTeeExec(va_dpy, protected_session, buffer);
+ *
+ * vaDestroyBuffer(va_dpy, buffer);
+ * \endcode
+ */
+
+/**
+ * ProtectedSessions and Contexts
+ *
+ * Context represents a "virtual" video decode, encode or video processing
+ * pipeline. Surfaces are render targets for a given context. The data in the
+ * surfaces are not accessible to the client except if derived image is supported
+ * and the internal data format of the surface is implementation specific.
+ *
+ * A protected session is a session between application and GPU hardware.
+ * Protected sessions can be attached to display to enable protection in the pipeline.
+ *
+ * Both contexts and protected sessions are identified by unique IDs and its
+ * implementation specific internals are kept opaque to the clients
+ */
+typedef VAContextID VAProtectedSessionID;
+
+/** \brief Input buffer of VAProtectedSessionTeeExecBuffer or VAProtectedSessionHwUpdateBuffer */
+typedef struct _VAProtectedFuncionBufferInput
+{
+    /** \brief Contains the size of the input data. */
+    uint32_t    data_size;
+    /** \brief The input data. */
+    void        *input_data;
+} VAProtectedFuncionBufferInput;
+
+/** \brief Output buffer of VAProtectedSessionTeeExecBuffer or VAProtectedSessionHwUpdateBuffer */
+typedef struct _VAProtectedFuncionBufferOutput
+{
+    /** \brief The maximum size of data that the driver can return in the output buffer. */
+    uint32_t    max_data_size;
+    /** \brief Returns the size of the output data written by the driver. */
+    uint32_t    data_size;
+    /** \brief Output data written by the driver. */
+    void        *output_data;
+} VAProtectedFuncionBufferOutput;
```

Comment:
I don't see the actual difference between input and output buffers. To me they can be just combined under single definition:
typedef struct _VAProtectedBuffer
{
    /** \brief Allocated space for the buffer. */
    uint32_t    max_data_size;
    /** \brief Size of the actual data inside the buffer. */
    uint32_t    data_size;
    /** \brief Data stored in the buffer. */
    void        *data;
} VAProtectedBuffer;

Reply:
There are cases that need both input and output buffer.